### PR TITLE
Buoyancy update in linear time.

### LINKED ISSUE #21: Generic vertical integral operator.
Should convert `∫δρgdz!(g::Grid, c::PlanetaryConstants, δρ::CellField, δρz::FaceFieldZ, pHY′::CellField)` to a generic vertical integral operator. Is this one particularly special?

### LINKED ISSUE #169: Vertical integrals should take linear time, not quadratic.
Thanks to some preliminary CSI profiling by @banex19 we have some timings for the time-stepping loop
```
Function time_step!: 1 times (average runtime : 20186 ms)
Function update_buoyancy!: 1 times (average runtime : 8631 ms)
Function calculate_interior_source_terms!: 1 times (average runtime : 4315 ms)
Function solve_poisson_3d_ppn_planned!: 1 times (average runtime : 1131 ms)
Function update_velocities_and_tracers!: 1 times (average runtime : 728 ms)
Function adams_bashforth_update_source_terms!: 1 times (average runtime : 526 ms)
Function store_previous_source_terms!: 1 times (average runtime : 353 ms)
Function calculate_source_term_divergence!: 1 times (average runtime : 295 ms)
```

I kind of overlooked this but the `update_buoyancy!` function is taking up 43% of the time stepping cycles because each grid point gets one thread that calculates the buoyancy for all grid points above them (quadratic time for no reason).

Well, the reason I did it this way was to avoid having to synchronize between threads, but there is no need to synchronize. This should be changed so that each column gets a thread that is responsible for calculating the buoyancy in that column. So when launching this kernel with CUDA it should be called with `threads=(Tx, Ty), blocks=(Bx, By)`. This is probably only performant for large grids with `Nx*Ny >> 1024`.

On the CPU this might speed up `update_buoyancy!` by ~10x and the entire time stepping by ~60%! Should speed up the GPU models as well.

Note: The absolute times themselves don't mean much as the profiler is pretty sophisticated/invasive.

### PR DESCRIPTION
**Pull request by ali-ramadhan**
Resolves #21
Resolves #169

### CODE REVIEW

**Comment by ali-ramadhan** [comment]
Wow this helped much more than I thought it would! Compared with the last time we benchmarked in https://github.com/climate-machine/Oceananigans.jl/pull/147#issuecomment-479474578
* the 256^3 Float64 CPU model is 5.3x faster!
* the 256^3 Float64 GPU model is 4.8x faster!

Thank you @banex19!

cc: @sandreza 

---
```
 ──────────────────────────────────────────────────────────────────────────────────────────────────
             Oceananigans.jl benchmarks                    Time                   Allocations      
                                                   ──────────────────────   ───────────────────────
                 Tot / % measured:                       227s / 45.6%           18.7GiB / 0.06%    

 Section                                   ncalls     time   %tot     avg     alloc   %tot      avg
 ──────────────────────────────────────────────────────────────────────────────────────────────────
 256x256x256 static ocean (CPU, Float32)       10    54.4s  52.5%   5.44s   60.0KiB  0.48%  6.00KiB
 256x256x256 static ocean (CPU, Float64)       10    36.9s  35.6%   3.69s   77.8KiB  0.62%  7.78KiB
 128x128x128 static ocean (CPU, Float32)       10    6.38s  6.16%   638ms   60.0KiB  0.48%  6.00KiB
 128x128x128 static ocean (CPU, Float64)       10    4.04s  3.90%   404ms   77.8KiB  0.62%  7.78KiB
  64x 64x 64 static ocean (CPU, Float32)       10    748ms  0.72%  74.8ms   60.0KiB  0.48%  6.00KiB
  64x 64x 64 static ocean (CPU, Float64)       10    412ms  0.40%  41.2ms   77.8KiB  0.62%  7.78KiB
 256x256x256 static ocean (GPU, Float64)       10    284ms  0.27%  28.4ms   1.59MiB  12.9%   163KiB
 256x256x256 static ocean (GPU, Float32)       10    243ms  0.23%  24.3ms   1.35MiB  11.0%   139KiB
  32x 32x 32 static ocean (CPU, Float32)       10   80.3ms  0.08%  8.03ms   60.0KiB  0.48%  6.00KiB
  32x 32x 32 static ocean (CPU, Float64)       10   45.2ms  0.04%  4.52ms   77.8KiB  0.62%  7.78KiB
 128x128x128 static ocean (GPU, Float64)       10   35.9ms  0.03%  3.59ms   1.59MiB  12.9%   163KiB
 128x128x128 static ocean (GPU, Float32)       10   32.3ms  0.03%  3.23ms   1.35MiB  11.0%   139KiB
  64x 64x 64 static ocean (GPU, Float64)       10   6.54ms  0.01%   654μs   1.59MiB  12.9%   163KiB
  64x 64x 64 static ocean (GPU, Float32)       10   6.14ms  0.01%   614μs   1.35MiB  11.0%   139KiB
  32x 32x 32 static ocean (GPU, Float64)       10   5.77ms  0.01%   577μs   1.59MiB  12.9%   163KiB
  32x 32x 32 static ocean (GPU, Float32)       10   5.68ms  0.01%   568μs   1.35MiB  11.0%   139KiB
 ──────────────────────────────────────────────────────────────────────────────────────────────────

CPU Float64 -> Float32 speedup:
 32x 32x 32 static ocean: 0.562
 64x 64x 64 static ocean: 0.551
128x128x128 static ocean: 0.633
256x256x256 static ocean: 0.677

GPU Float64 -> Float32 speedup:
 32x 32x 32 static ocean: 1.015
 64x 64x 64 static ocean: 1.066
128x128x128 static ocean: 1.114
256x256x256 static ocean: 1.167

CPU -> GPU speedup:
 32x 32x 32 static ocean (Float32): 14.138
 32x 32x 32 static ocean (Float64): 7.829
 64x 64x 64 static ocean (Float32): 121.806
 64x 64x 64 static ocean (Float64): 62.924
128x128x128 static ocean (Float32): 197.906
128x128x128 static ocean (Float64): 112.417
256x256x256 static ocean (Float32): 223.748
256x256x256 static ocean (Float64): 129.923
```

**Comment by banex19** [comment]
That's amazing, I'm glad I could help. I'll run some more profiling in the near future.

File: src/time_steppers.jl [hunk 1]
```diff
@@ -120,23 +119,15 @@ function store_previous_source_terms!(grid::Grid, Gu, Gv, Gw, GT, GS, Gpu, Gpv,
...
-
-                ∫δρ = (-ρ₀*βT*(T[i, j, 1]-T₀))
-                for k′ in 2:k
-                    ∫δρ += ((-ρ₀*βT*(T[i, j, k′-1]-T₀)) + (-ρ₀*βT*(T[i, j, k′]-T₀)))
-                end
-                @inbounds pHY′[i, j, k] = 0.5 * gΔz * ∫δρ
+    @loop for j in (1:grid.Ny; (blockIdx().y - 1) * blockDim().y + threadIdx().y)
+        @loop for i in (1:grid.Nx; (blockIdx().x - 1) * blockDim().x + threadIdx().x)
+            @inbounds pHY′[i, j, 1] = 0.5 * gΔz * δρ(eos, T, i, j, 1)
+            for k in 2:grid.Nz
```
Comment (vchuravy) [inline src/time_steppers.jl]: You might want to use the `@unroll` macro from GPUifyLoops here. Is `grid.Nz` a compile time constant value?

Comment (ali-ramadhan) [reply src/time_steppers.jl]: Ah yes, `grid.Nz` should be a compile time constant `::Int`. Didn't know about `@unroll`, will give it a try!

**Comment by ali-ramadhan** [comment]
@vchuravy Just benchmarked again with `@unroll` but couldn't see any difference in performance.

Good to know about it though, will try it if we have more loops that could be unrolled.

**Comment by vchuravy** [comment]
LLVM might have decided to do it anyway, but since GPUs tend to prefer straightline code it is a useful tool to have access to.

**Pull request merged**